Awesome Books Project - Refactor

lmunozm1702 opened the pull request:
## Re-create the project as required:

- Use ES6 syntax
- Refactor the original JS code with the use of modules.
- Change original methods and functions to use the arrow function.
- Luxon module utilization to improve Date & Time management.

### I include some modifications in the styles sheet to improve the presentation.

lmunozm1702 committed:
Refactor to ES6 syntax and modules

lmunozm1702 committed:
Refactor date & time using luxon module.

Add luxon module (luxon.js) from https://github.com/moment/luxon/

update setCurrentTime.js to use luxon module.

lmunozm1702 committed:
Add CSS styles

Update index.css, index.html, setNavBar.js and awesomeBooks.js to improve the style on views.

lmunozm1702 committed:
Add README.md

lmunozm1702 committed:
Fix linters issues

lmunozm1702 commented:
The ESLint errors are in Luxon module (3.084 errors). I left the errors because are in the external code asked to use in the requirements. 

The instructions say I can't modify the linter configuration file to avoid this file.

![image](https://user-images.githubusercontent.com/53025062/198994816-236cad0c-4c7e-434c-96f7-0d02722502bd.png)

ha-manel reviewed (commented):
# STATUS: INVALID 🔴 
Hi @lmunozm1702 
I'm sure you put a lot of effort into this project but this PR is invalid because you have failed linters checks😓 

### Even if the errors are in luxon you can always fix those errors (most of them are about missing semicolons and the use of single quotes, so it won't affect the code in luxon).

Kindly fix the issue and submit the PR for review again 👍 
Best of luck!

------
_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

lmunozm1702 commented:
HI @ha-manel I fix about 2.000 issues with missing semicolons and double quotes like you say, in the Luxor library. But I have yet about 1.000 more linter errors into the library, I am in the 2nd. week on module 2 and don't understand some issues to fix them, and I have just 4 hours to do the project and I feel I need 8 more to fix the library and test all the changes... I go to ask to student success if I can take today to make the fixing and if they can extend my week if I need more time on Saturday.

The resume of pending errors into the library are:

<google-sheets-html-origin><style type="text/css"><!--br {mso-data-placement:same-cell;}--></style>

Expected this semicolon to be at the end of the previous line | 154
-- | --
Expected 'undefined' and instead saw 'void' | 130
Unexpected unnamed function | 34
This line has a length of 101. Maximum allowed is 100 | 30
Unexpected var, use let or const instead | 24
Unary operator '++' used | 24
All 'var' declarations must be at the top of the function scope | 24
Use the rest parameters instead of 'arguments' | 19
DateTime' was used before it was defined | 16
This line has a length of 108. Maximum allowed is 100 | 15
This line has a length of 103. Maximum allowed is 100 | 15
This line has a length of 109. Maximum allowed is 100 | 13
This line has a length of 106. Maximum allowed is 100 | 13
Unexpected dangling '_' in '_key' | 12
Unexpected dangling '_' in '_proto' | 11
This line has a length of 110. Maximum allowed is 100 | 11
This line has a length of 102. Maximum allowed is 100 | 11
This line has a length of 111. Maximum allowed is 100 | 10
Unexpected dangling '_' in '_key2' | 9
This line has a length of 119. Maximum allowed is 100 | 9
Unexpected mix of '\|\|' and '&&'. Use parentheses to clarify the intended order of operations | 8
Unexpected mix of '&&' and '\|\|'. Use parentheses to clarify the intended order of operations | 8
Unexpected mix of '*' and '/'. Use parentheses to clarify the intended order of operations | 8
Unexpected assignment within a 'for' statement | 8
This line has a length of 114. Maximum allowed is 100 | 8
This line has a length of 104. Maximum allowed is 100 | 8
Locale' was used before it was defined | 8
Unexpected dangling '_' in '_this' | 7
This line has a length of 140. Maximum allowed is 100 | 7
This line has a length of 127. Maximum allowed is 100 | 7
for..in loops iterate over the entire prototype chain, which is virtually never what you want. Use Object.{keys,values,entries}, and iterate over the resulting array | 7
Unexpected dangling '_' in '_i' | 6
This line has a length of 112. Maximum allowed is 100 | 6
IANAZone' was used before it was defined | 6
FixedOffsetZone' was used before it was defined | 6
Use array destructuring | 5
Unexpected dangling '_' in '_ref' | 5
This line has a length of 143. Maximum allowed is 100 | 5
This line has a length of 142. Maximum allowed is 100 | 5
This line has a length of 123. Maximum allowed is 100 | 5
This line has a length of 120. Maximum allowed is 100 | 5
This line has a length of 116. Maximum allowed is 100 | 5
This line has a length of 115. Maximum allowed is 100 | 5
Return statement should not contain assignment | 5
friendlyDateTime' was used before it was defined | 5
dateTimes' used outside of binding context | 5
units' used outside of binding context | 4
Unexpected dangling '_' in '_step' | 4
Unexpected dangling '_' in '_len' | 4
Unexpected dangling '_' in '_iterator' | 4
This line has a length of 171. Maximum allowed is 100 | 4
This line has a length of 163. Maximum allowed is 100 | 4
This line has a length of 138. Maximum allowed is 100 | 4
This line has a length of 134. Maximum allowed is 100 | 4
This line has a length of 132. Maximum allowed is 100 | 4
This line has a length of 130. Maximum allowed is 100 | 4
This line has a length of 129. Maximum allowed is 100 | 4
This line has a length of 121. Maximum allowed is 100 | 4
This line has a length of 118. Maximum allowed is 100 | 4
This line has a length of 113. Maximum allowed is 100 | 4
This line has a length of 107. Maximum allowed is 100 | 4
Formatter' was used before it was defined | 4
Do not nest ternary expressions | 4
Use the spread operator instead of '.apply()' | 3
Unexpected dangling '_' in '_this2' | 3
Unexpected dangling '_' in '_ref6' | 3
Unexpected dangling '_' in '_ref3' | 3
Unexpected dangling '_' in '_ref2' | 3
Unexpected dangling '_' in '_len2' | 3
Unexpected dangling '_' in '_key4' | 3
ts' is defined but never used | 3
This line has a length of 151. Maximum allowed is 100 | 3
This line has a length of 139. Maximum allowed is 100 | 3
This line has a length of 136. Maximum allowed is 100 | 3
This line has a length of 133. Maximum allowed is 100 | 3
This line has a length of 126. Maximum allowed is 100 | 3
This line has a length of 125. Maximum allowed is 100 | 3
This line has a length of 124. Maximum allowed is 100 | 3
This line has a length of 122. Maximum allowed is 100 | 3
Interval' was used before it was defined | 3
Function '_unsupportedIterableToArray' expected no return value | 3
Expected an assignment or function call and instead saw an expression | 3
_this' is never reassigned. Use 'const' instead | 3
Unnecessary escape character: \[ | 2
Unexpected use of continue statement | 2
Unexpected use of 'isNaN'. Use Number.isNaN instead https://github.com/airbnb/javascript#standard-library--isnan | 2
Unexpected mix of '-' and '%'. Use parentheses to clarify the intended order of operations | 2
Unexpected dangling '_' in '_zone' | 2
Unexpected dangling '_' in '_step3' | 2
Unexpected dangling '_' in '_step2' | 2
Unexpected dangling '_' in '_ref5' | 2
Unexpected dangling '_' in '_ref4' | 2
Unexpected dangling '_' in '_opts' | 2
Unexpected dangling '_' in '_options' | 2
Unexpected dangling '_' in '_key3' | 2
Unexpected dangling '_' in '_iterator3' | 2
Unexpected dangling '_' in '_iterator2' | 2
Unexpected dangling '_' in '_i2' | 2
Unexpected dangling '_' in '_Array$prototype' | 2
This line has a length of 211. Maximum allowed is 100 | 2
This line has a length of 170. Maximum allowed is 100 | 2
This line has a length of 166. Maximum allowed is 100 | 2
This line has a length of 160. Maximum allowed is 100 | 2
This line has a length of 149. Maximum allowed is 100 | 2
This line has a length of 141. Maximum allowed is 100 | 2
This line has a length of 137. Maximum allowed is 100 | 2
This line has a length of 117. Maximum allowed is 100 | 2
This line has a length of 105. Maximum allowed is 100 | 2
The body of a for-in should be wrapped in an if statement to filter unwanted properties from the prototype | 2
The '__proto__' property is deprecated | 2
keys' used outside of binding context | 2
Expected a default case | 2
_construct' is a function | 2
_arrayLikeToArray' was used before it was defined | 2
ZoneIsAbstractError' was used before it was defined | 1
Wrapper' was used before it was defined | 1
Unnecessary escape character: \^ | 1
Unnecessary escape character: \- | 1
Unexpected dangling '_' in '_wrapNativeSuper' | 1
Unexpected dangling '_' in '_unsupportedIterableToArray' | 1
Unexpected dangling '_' in '_uncomputeOrdinal2' | 1
Unexpected dangling '_' in '_uncomputeOrdinal' | 1
Unexpected dangling '_' in '_toISOTime' | 1
Unexpected dangling '_' in '_toISODate' | 1
Unexpected dangling '_' in '_this4' | 1
Unexpected dangling '_' in '_this3' | 1
Unexpected dangling '_' in '_this$plus' | 1
Unexpected dangling '_' in '_split' | 1
Unexpected dangling '_' in '_setPrototypeOf' | 1
Unexpected dangling '_' in '_ref7$includeZone' | 1
Unexpected dangling '_' in '_ref7$includeOffsetSp' | 1
Unexpected dangling '_' in '_ref7$includeOffset' | 1
Unexpected dangling '_' in '_ref7' | 1
Unexpected dangling '_' in '_ref6$suppressSeconds' | 1
Unexpected dangling '_' in '_ref6$suppressMillise' | 1
Unexpected dangling '_' in '_ref6$locale' | 1
Unexpected dangling '_' in '_ref6$includePrefix' | 1
Unexpected dangling '_' in '_ref6$includeOffset' | 1
Unexpected dangling '_' in '_ref6$format' | 1
Unexpected dangling '_' in '_ref6$extendedZone' | 1
Unexpected dangling '_' in '_ref5$locale' | 1
Unexpected dangling '_' in '_ref5$format' | 1
Unexpected dangling '_' in '_ref4$suppressSeconds' | 1
Unexpected dangling '_' in '_ref4$suppressMillise' | 1
Unexpected dangling '_' in '_ref4$numberingSystem' | 1
Unexpected dangling '_' in '_ref4$locObj' | 1
Unexpected dangling '_' in '_ref4$locale' | 1
Unexpected dangling '_' in '_ref4$includeOffset' | 1
Unexpected dangling '_' in '_ref4$format' | 1
Unexpected dangling '_' in '_ref4$extendedZone' | 1
Unexpected dangling '_' in '_ref3$separator' | 1
Unexpected dangling '_' in '_ref3$numberingSystem' | 1
Unexpected dangling '_' in '_ref3$locObj' | 1
Unexpected dangling '_' in '_ref3$locale' | 1
Unexpected dangling '_' in '_ref2$outputCalendar' | 1
Unexpected dangling '_' in '_ref2$numberingSystem' | 1
Unexpected dangling '_' in '_ref2$locObj' | 1
Unexpected dangling '_' in '_ref2$locale' | 1
Unexpected dangling '_' in '_ref2$keepLocalTime' | 1
Unexpected dangling '_' in '_ref2$keepCalendarTim' | 1
Unexpected dangling '_' in '_ref$outputCalendar' | 1
Unexpected dangling '_' in '_ref$numberingSystem' | 1
Unexpected dangling '_' in '_ref$locObj' | 1
Unexpected dangling '_' in '_ref$locale' | 1
Unexpected dangling '_' in '_proto4' | 1
Unexpected dangling '_' in '_proto3' | 1
Unexpected dangling '_' in '_proto2' | 1
Unexpected dangling '_' in '_patterns$_i' | 1
Unexpected dangling '_' in '_patterns' | 1
Unexpected dangling '_' in '_parseSQL' | 1
Unexpected dangling '_' in '_parseRFC2822Date' | 1
Unexpected dangling '_' in '_parseLocaleString' | 1
Unexpected dangling '_' in '_parseISOTimeOnly' | 1
Unexpected dangling '_' in '_parseISODuration' | 1
Unexpected dangling '_' in '_parseISODate' | 1
Unexpected dangling '_' in '_parseHTTPDate' | 1
Unexpected dangling '_' in '_parseFromTokens' | 1
Unexpected dangling '_' in '_opts$numberingSystem' | 1
Unexpected dangling '_' in '_opts$locale' | 1
Unexpected dangling '_' in '_options$numberingSys' | 1
Unexpected dangling '_' in '_options$locale' | 1
Unexpected dangling '_' in '_objToTS4' | 1
Unexpected dangling '_' in '_objToTS3' | 1
Unexpected dangling '_' in '_objToTS2' | 1
Unexpected dangling '_' in '_objToTS' | 1
Unexpected dangling '_' in '_objectWithoutPropertiesLoose' | 1
Unexpected dangling '_' in '_Object$keys2' | 1
Unexpected dangling '_' in '_Object$keys' | 1
Unexpected dangling '_' in '_numberingSystemsUTF' | 1
Unexpected dangling '_' in '_match' | 1
Unexpected dangling '_' in '_macroTokenToFormatOpts' | 1
Unexpected dangling '_' in '_len4' | 1
Unexpected dangling '_' in '_len3' | 1
Unexpected dangling '_' in '_lastOpts2' | 1
Unexpected dangling '_' in '_lastOpts' | 1
Unexpected dangling '_' in '_isNativeReflectConstruct' | 1
Unexpected dangling '_' in '_isNativeFunction' | 1
Unexpected dangling '_' in '_intervals$sort$reduc' | 1
Unexpected dangling '_' in '_inheritsLoose' | 1
Unexpected dangling '_' in '_highOrderDiffs' | 1
Unexpected dangling '_' in '_getPrototypeOf' | 1
Unexpected dangling '_' in '_Formatter$create$res' | 1
Unexpected dangling '_' in '_formatted$i' | 1
Unexpected dangling '_' in '_fixOffset' | 1
Unexpected dangling '_' in '_fixed' | 1
Unexpected dangling '_' in '_extends' | 1
Unexpected dangling '_' in '_explainFromTokens' | 1
Unexpected dangling '_' in '_excluded2' | 1
Unexpected dangling '_' in '_excluded' | 1
Unexpected dangling '_' in '_ex' | 1
Unexpected dangling '_' in '_Duration$fromMillis' | 1
Unexpected dangling '_' in '_dur' | 1
Unexpected dangling '_' in '_differs$_i' | 1
Unexpected dangling '_' in '_differs' | 1
Unexpected dangling '_' in '_diff' | 1
Unexpected dangling '_' in '_defineProperties' | 1
Unexpected dangling '_' in '_cursor$plus3' | 1
Unexpected dangling '_' in '_cursor$plus2' | 1
Unexpected dangling '_' in '_cursor$plus' | 1
Unexpected dangling '_' in '_createForOfIteratorHelperLoose' | 1
Unexpected dangling '_' in '_createClass' | 1
Unexpected dangling '_' in '_construct' | 1
Unexpected dangling '_' in '_cache' | 1
Unexpected dangling '_' in '_buildRegex' | 1
Unexpected dangling '_' in '_arrayLikeToArray' | 1
Unexpected assignment within an 'if' statement | 1
This line has a length of 477. Maximum allowed is 100 | 1
This line has a length of 464. Maximum allowed is 100 | 1
This line has a length of 415. Maximum allowed is 100 | 1
This line has a length of 364. Maximum allowed is 100 | 1
This line has a length of 343. Maximum allowed is 100 | 1
This line has a length of 335. Maximum allowed is 100 | 1
This line has a length of 311. Maximum allowed is 100 | 1
This line has a length of 298. Maximum allowed is 100 | 1
This line has a length of 288. Maximum allowed is 100 | 1
This line has a length of 282. Maximum allowed is 100 | 1
This line has a length of 270. Maximum allowed is 100 | 1
This line has a length of 257. Maximum allowed is 100 | 1
This line has a length of 252. Maximum allowed is 100 | 1
This line has a length of 251. Maximum allowed is 100 | 1
This line has a length of 249. Maximum allowed is 100 | 1
This line has a length of 243. Maximum allowed is 100 | 1
This line has a length of 241. Maximum allowed is 100 | 1
This line has a length of 240. Maximum allowed is 100 | 1
This line has a length of 226. Maximum allowed is 100 | 1
This line has a length of 225. Maximum allowed is 100 | 1
This line has a length of 222. Maximum allowed is 100 | 1
This line has a length of 213. Maximum allowed is 100 | 1
This line has a length of 205. Maximum allowed is 100 | 1
This line has a length of 199. Maximum allowed is 100 | 1
This line has a length of 189. Maximum allowed is 100 | 1
This line has a length of 184. Maximum allowed is 100 | 1
This line has a length of 182. Maximum allowed is 100 | 1
This line has a length of 181. Maximum allowed is 100 | 1
This line has a length of 178. Maximum allowed is 100 | 1
This line has a length of 177. Maximum allowed is 100 | 1
This line has a length of 175. Maximum allowed is 100 | 1
This line has a length of 173. Maximum allowed is 100 | 1
This line has a length of 169. Maximum allowed is 100 | 1
This line has a length of 167. Maximum allowed is 100 | 1
This line has a length of 164. Maximum allowed is 100 | 1
This line has a length of 159. Maximum allowed is 100 | 1
This line has a length of 157. Maximum allowed is 100 | 1
This line has a length of 156. Maximum allowed is 100 | 1
This line has a length of 154. Maximum allowed is 100 | 1
This line has a length of 152. Maximum allowed is 100 | 1
This line has a length of 148. Maximum allowed is 100 | 1
This line has a length of 145. Maximum allowed is 100 | 1
This line has a length of 144. Maximum allowed is 100 | 1
This line has a length of 135. Maximum allowed is 100 | 1
This line has a length of 131. Maximum allowed is 100 | 1
This line has a length of 128. Maximum allowed is 100 | 1
SystemZone' was used before it was defined | 1
s, 34 warnings) | 1
regexes' used outside of binding context | 1
patterns' used outside of binding context | 1
Parent' is defined but never used | 1
otherZone' is defined but never used | 1
opts' is defined but never used | 1
Opening curly brace does not appear on the same line as controlling statement | 1
Math.pow' is restricted from being used. Use the exponentiation operator (**) instead | 1
LuxonError' was used before it was defined | 1
luxon' is assigned a value but never used | 1
InvalidZone' was used before it was defined | 1
InvalidUnitError' was used before it was defined | 1
InvalidIntervalError' was used before it was defined | 1
InvalidDurationError' was used before it was defined | 1
InvalidDateTimeError' was used before it was defined | 1
InvalidArgumentError' was used before it was defined | 1
intervals' used outside of binding context | 1
Identifier '_patterns$_i' is not in camel case | 1
Identifier '_differs$_i' is not in camel case | 1
Function declared in a loop contains unsafe references to variable(s) 'idx' | 1
formatOptsToTokens' was used before it was defined | 1
format' is defined but never used | 1
extractors' used outside of binding context | 1
Duration' was used before it was defined | 1
ConflictingSpecificationError' was used before it was defined | 1
Class' is defined but never used | 1
arr2' used outside of binding context | 1
args' is defined but never used | 1
_wrapNativeSuper' is a function | 1
_setPrototypeOf' was used before it was defined | 1
_setPrototypeOf' is a function | 1
_getPrototypeOf' is a function | 1
_extends' is a function | 1

lmunozm1702 commented:
When I used --fix to solve the first 2.000 issues, the library just don't work more. I need to do a rollback to get the functionality again.

lmunozm1702 commented:
And the instructions of the project don't permit the installation of the library with npm...

![image](https://user-images.githubusercontent.com/53025062/199025633-55ee732e-6f9f-47d3-b4c1-d007dd568df6.png)

I don't know how to go forward.

lmunozm1702 committed:
Add linter disable in 3rd party library.

ha-manel commented:
Hey @lmunozm1702 since this is a 3rd party library and fixing linters changes its behavior I think you can disable linters for that file 👍 Best of luck 💯

lmunozm1702 commented:
Thanks, @ha-manel I do like you say, and I pass the linter validations!

youmari commented on assets/js/modules/awesomebooks.js:
```diff
@@ -0,0 +1,55 @@
+    this.id = 0;
+  }
+
+  renderBooks() {
+    this.awesomeBooks2.forEach((book) => {
+      this.addBook(book, '.books-content');
+    });
+  }
+
+  addBook(Book, booksDiv) {
+    this.id += 1;
+    Book.id = this.id;
+
+    this.awesomeBooks.push(Book);
+
+    const singleBook = document.createElement('div');
+    singleBook.classList.add('single-book');
+    singleBook.id = `Book${Book.id}`;
+    document.querySelector(booksDiv).appendChild(singleBook);
+
+    const titleDiv = document.createElement('div');
+    titleDiv.textContent = `'${Book.title}' by  ${Book.author}`;
+    singleBook.appendChild(titleDiv);
+
+    const buttonDiv = document.createElement('i');
+    buttonDiv.className = ('fa-solid fa-trash-can');
+    buttonDiv.id = `button_${Book.id}`;
+    buttonDiv.onclick = () => {
+      this.removeBook(Book.id);
+    };
+    singleBook.appendChild(buttonDiv);
+    this.addToLocalStorage();
+  }
+
+  addToLocalStorage() {
+    localStorage.setItem('awesomeBooks', JSON.stringify(this.awesomeBooks));
+  }
+
```
- [x] it would be better if you turn these methods to `arrow functions` in order to use the es6 features and also match the requirements

lmunozm1702 replied:
It's done!, I redefine the 4 methods to use arrow function.

youmari reviewed (changes requested):
Hi @lmunozm1702 , 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!


## Highlights
✔️ Great commit messages
✔️ PRO PR
✔️ No linter error 

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

lmunozm1702 committed:
Fix code review issues

Update awesomebooks.js to define renderBooks, addBook, addToLocalStorage and removeBook using arrow function.

leonardodiasb reviewed (approved):
# Status: ✔️ Project Approved ✔️ 

Hello @lmunozm1702 🔥 

You worked hard in all the instructions that the previous Code Reviewers give you👍🏽 

Now you can merge this branch 🚀 

## Highlights

 - Good use of arrow functions. ✔️ 
 - Correct use of Github flow. ✔️ 
 - No linter errors. ✔️ 

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
